### PR TITLE
Add FAQ entry for viewing trace exceptions in Docs

### PR DESCRIPTION
**Pull request by BenWilson2**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/BenWilson2/mlflow/pull/12309?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12309/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12309
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Add FAQ for trace exception viewing

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [X] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [X] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [X] Instructions

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [X] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [X] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [X] No (this PR will be included in the next minor release)

### CODE REVIEW

File: docs/source/llms/tracing/index.rst [hunk 1]
```diff
@@ -690,3 +690,85 @@ Client API
...
+a trace from a previous session, you must use the ``search_traces`` client API (shown below).
+
+
+.. code-block:: python
+
+    trace = mlflow.get_trace(span.request_id)
+    trace_data_spans = trace.data.spans
+    trace_data.to_dict()
+
+In an interactive environment, such as a Jupyter Notebook, ``stdout`` will render an output like this:
```
Comment (harupy) [inline docs/source/llms/tracing/index.rst:729]:
```suggestion
    print(trace_data.to_dict())

Output:
```

can we use `print`?

File: docs/source/llms/tracing/index.rst [hunk 2]
```diff
@@ -690,3 +690,85 @@ Client API
...
+                                     line 4, in <cell line: 1>\n    
+                                        raise Exception("Intentionally throwing!")\nException: Intentionally throwing!\n',
+            'exception.escaped': 'False'}}]}],
+        'request': '{"input": "Exception should log as event"}',
+        'response': None}
+
+The ``exception.stacktrace`` attribute contains the full stack trace of the Exception that was raised during the span's execution.
+
+Alternatively, if you were to use the MLflowClient API to search traces, the access to retrieve the span's event data from the failure would be 
+slightly different (due to the return value being a ``Pandas`` DataFrame). To use the ``search_traces`` API to access the same exception data would
```
Comment (harupy) [inline docs/source/llms/tracing/index.rst:764]:
```suggestion
slightly different (due to the return value being a ``pandas`` DataFrame). To use the ``search_traces`` API to access the same exception data would 
```

File: docs/source/llms/tracing/index.rst (hunk 1 above)
Comment (BenWilson2) [reply docs/source/llms/tracing/index.rst:729]: print collapses the output into a string-wrapped blob that is impossible to read, whereas the Jupyter renderer will pretty print dictionaries. 
I'm fine with adding the print, I just left it out to make it more visually appealing for users who copy the code to test it out

Comment (BenWilson2) [reply docs/source/llms/tracing/index.rst:729]: Screenshot 2024-06-10 at 10 07 13 PM
(comparison)

File: docs/source/llms/tracing/index.rst [hunk 3]
```diff
@@ -690,3 +690,85 @@ Client API
...
+
+    with mlflow.start_span(name="A Problematic Span") as span:
+        span.set_inputs({"input": "Exception should log as event"})
+        span.set_attribute("a", "b")
+        raise Exception("Intentionally throwing!")
+        span.set_outputs({"This": "should not be recorded"})
+
+When running this, an Exception will be thrown, as expected. However, a trace is still logged to the active experiment and can be retrieved as follows:
+
+Note that the ``get_trace`` API is **only available** for retrieving traces from the current active session. If you are attempting to retrieve
```
Comment (B-Step62) [inline docs/source/llms/tracing/index.rst:719]: Actually it won't take 10 mins to support backend fetch in `get_trace()`. Shall we do that and remove this warning?

File: docs/source/llms/tracing/index.rst (hunk 1 above)
Comment (harupy) [reply docs/source/llms/tracing/index.rst:729]: @BenWilson2 

can we use `trace_data.to_json(pretty=True)`?

File: docs/source/llms/tracing/index.rst [hunk 4]
```diff
@@ -690,3 +690,85 @@ Client API
...
+
+The ``exception.stacktrace`` attribute contains the full stack trace of the Exception that was raised during the span's execution.
+
+Alternatively, if you were to use the MLflowClient API to search traces, the access to retrieve the span's event data from the failure would be 
+slightly different (due to the return value being a ``pandas`` DataFrame). To use the ``search_traces`` API to access the same exception data would 
+be as follows:
+
+.. code-block:: python
+
+    traces = mlflow.search_traces(
```
Comment (B-Step62) [inline docs/source/llms/tracing/index.rst:769]: Shouldn't we use `mlflow.MlflowClient().search_traces` here?

File: docs/source/llms/tracing/index.rst [hunk 5]
```diff
@@ -690,3 +690,85 @@ Client API
...
+            'mlflow.spanInputs': '{"input": "Exception should log as event"}',
+            'a': '"b"'},
+        'events': [{'name': 'exception',
+            'timestamp': 1718063629190527000,
+            'attributes': {'exception.type': 'Exception',
+            'exception.message': 'Intentionally throwing!',
+            'exception.stacktrace': 'Traceback (most recent call last):\n  
+                                     File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/opentelemetry/trace/__init__.py", 
+                                     line 573, in use_span\n    
+                                        yield span\n  File "/Users/benjamin.wilson/repos/mlflow-fork/mlflow/mlflow/tracing/fluent.py",
```
Comment (harupy) [inline docs/source/llms/tracing/index.rst:752]:
```suggestion
                                     File "/usr/local/lib/python3.8/site-packages/opentelemetry/trace/__init__.py", 
                                     line 573, in use_span\n    
                                        yield span\n  File "/usr/local/lib/python3.8/site-package/mlflow/mlflow/tracing/fluent.py", 
```

File: docs/source/llms/tracing/index.rst (hunk 3 above)
Comment (BenWilson2) [reply docs/source/llms/tracing/index.rst:719]: +100

File: docs/source/llms/tracing/index.rst (hunk 1 above)
Comment (BenWilson2) [reply docs/source/llms/tracing/index.rst:729]: TraceData doesn't have a `to_json` method

File: docs/source/llms/tracing/index.rst (hunk 4 above)
Comment (BenWilson2) [reply docs/source/llms/tracing/index.rst:769]: I like to make up APIs as I go when I manually copy from one place to anot- 
Yes, yes we should. ;) Good catch

File: docs/source/llms/tracing/index.rst (hunk 5 above)
Comment (BenWilson2) [reply docs/source/llms/tracing/index.rst:752]: +1

File: docs/source/llms/tracing/index.rst [hunk 6]
```diff
@@ -690,3 +690,85 @@ Client API
...
+
+Alternatively, if you were to use the MLflowClient API to search traces, the access to retrieve the span's event data from the failure would be 
+slightly different (due to the return value being a ``pandas`` DataFrame). To use the ``search_traces`` API to access the same exception data would 
+be as follows:
+
+.. code-block:: python
+
+    traces = mlflow.search_traces(
+        experiment_ids=[experiment.experiment_id]
+    )  # This returns a Pandas DataFrame
```
Comment (harupy) [inline docs/source/llms/tracing/index.rst:771]:
```suggestion
    )  # This returns a pandas DataFrame
```

File: docs/source/llms/tracing/index.rst [hunk 7]
```diff
@@ -690,3 +690,85 @@ Client API
...
+
+Note that the ``get_trace`` API is **only available** for retrieving traces from the current active session. If you are attempting to retrieve 
+a trace from a previous session, you must use the ``search_traces`` client API (shown below).
+
+
+.. code-block:: python
+
+    trace = mlflow.get_trace(span.request_id)
+    trace_data_spans = trace.data.spans
+    trace_data.to_dict()
```
Comment (harupy) [inline docs/source/llms/tracing/index.rst:727]:
```suggestion
    trace = mlflow.get_trace(span.request_id)
    trace_data_spans = trace.data.spans
    trace_data_spans.to_dict()
```

File: docs/source/llms/tracing/index.rst (hunk 1 above)
Comment (harupy) [reply docs/source/llms/tracing/index.rst:729]: Got it, then I'm fine with the current code.

File: docs/source/llms/tracing/index.rst (hunk 7 above)
Comment (BenWilson2) [reply docs/source/llms/tracing/index.rst]: another typo from testing. Ran the code blocks locally and everything's good now.

File: docs/source/llms/tracing/index.rst (hunk 1 above)
Comment (BenWilson2) [reply docs/source/llms/tracing/index.rst]: I actually just changed it to pprint. I think it's pretty close to the `Output` block in Jupyter, but now people running in command line can use it :)

File: docs/source/llms/tracing/index.rst (hunk 3 above)
Comment (B-Step62) [reply docs/source/llms/tracing/index.rst]: let me submit a PR in 10 mins🙂

**Review by harupy — APPROVED** [review]
LGTM!

File: docs/source/llms/tracing/index.rst (hunk 3 above)
Comment (B-Step62) [reply docs/source/llms/tracing/index.rst]: Filed:) https://github.com/mlflow/mlflow/pull/12311

**Pull request merged**